### PR TITLE
Data source for packer registry

### PR DESCRIPTION
**Pull request by azr**
### :hammer_and_wrench: Description

This add a new data source block that will display data from the Packer Registry.

⚠️ I also had to fix a few references for consul things after updating the dependency.

### :ship: Release Note
Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-hcp/blob/main/CHANGELOG.md):

```release-note
* provider: add data source for packer registry
```

### :building_construction: Acceptance tests

- [x] Are there any feature flags that are required to use this functionality? > Nope !
- [x] Have you added an acceptance test for the functionality being added?
- [x] Have you run the acceptance tests on this branch?

Output from acceptance testing:

```shellsession
$ make testacc TESTARGS='-run=TestAcc_dataSourcePacker'      
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/... -v -run=TestAcc_dataSourcePacker -timeout 120m
?       github.com/hashicorp/terraform-provider-hcp/internal/clients    [no test files]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-hcp/internal/consul     2.709s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-hcp/internal/input      3.274s [no tests to run]
=== RUN   TestAcc_dataSourcePacker
--- PASS: TestAcc_dataSourcePacker (9.86s)
PASS
ok      github.com/hashicorp/terraform-provider-hcp/internal/provider   10.540s

...
```

### CODE REVIEW

**Comment by azr** [comment]
Can someone tell me against what tier the acceptance tests are running ? Here or in private msg 🙂 !

**Review by SwampDragons — APPROVED** [review]
This looks good to me, though I haven't had a chance to test it. How are resources for this provider normally documented? I don't see any doc elements added here, but it's possible they're generated somewhere else and I'm just missing context.

File: internal/provider/data_source_packer_image.go
Comment (bcmdarroch) [inline internal/provider/data_source_packer_image.go:67]: Would it be possible to flatten this object at all?

**Review by bcmdarroch — COMMENTED** [review]
Looking good! Had a few questions and comments. Also, you can generate docs by running `go generate ./...`. Not sure why the PR checks didn't catch that 💭

File: internal/clients/hvn_route.go
Comment (azr) [reply internal/clients/hvn_route.go:123]: Gotcha, I need the features of this branch here : https://github.com/hashicorp/hcp-sdk-go/pull/29 that was created by @nywilken. Maybe we can wait for the sdk PR to be merged before merging this one.

File: internal/provider/data_source_packer_image.go
Comment (azr) [reply internal/provider/data_source_packer_image.go:27]: Got it, making it `bucket_id`

**Comment by azr** [comment]
Hello there, thanks for reviewing:

* [ ] make sure https://github.com/hashicorp/hcp-sdk-go/pull/29 is merged
* [ ] update hcp-sdk-go
* [ ] rename `bucket` to `bucket_id`
* [ ] put cloud-specific object into named/predefined objects

File: internal/provider/data_source_packer_image.go
Comment (azr) [reply internal/provider/data_source_packer_image.go:27]: Actually, thinking about it, this is a `slug`. Like `alpine`. So it is different in the sense that this is not an id, changing the dscription/docs message to remove the notion of `ID`.

Tell me what you think.

**Review by nywilken — APPROVED** [review]
This looks really good.

File: internal/clients/hvn_route.go
Comment (bcmdarroch) [reply internal/clients/hvn_route.go]: Yeah that sounds good.

File: internal/provider/data_source_packer_image.go
Comment (bcmdarroch) [reply internal/provider/data_source_packer_image.go]: I would still lean towards `bucket_id`. The rest of our resources also use slugs IDs, so that would be consistent.

File: internal/provider/data_source_packer_image_iteration.go
Comment (bcmdarroch) [inline internal/provider/data_source_packer_image_iteration.go:20]:
```suggestion
		Description: "The Packer Image data source iteration gets the most recent iteration (or build) of an image, given a channel.",
```

File: internal/provider/data_source_packer_image_iteration.go
Comment (bcmdarroch) [inline internal/provider/data_source_packer_image_iteration.go:15]: [nit] could drop this or leave a TODO.

File: internal/provider/data_source_packer_image.go
Comment (azr) [reply internal/provider/data_source_packer_image.go]: Its definite format will evolve for a bit. Almost there. I'll try to flatten it as much as I can. 🙂

File: internal/provider/data_source_packer_image_iteration.go
Comment (azr) [reply internal/provider/data_source_packer_image_iteration.go:20]: Thanks !

File: internal/provider/data_source_packer_image_iteration.go
Comment (azr) [reply internal/provider/data_source_packer_image_iteration.go:15]: Got it, yeah dropping it, I think 1 min is okay for now 🙂 
```suggestion
```

**Comment by azr** [comment]
With the following snippet:

```hcl
terraform {
  required_providers {
    hcp = {
      source = "localhost/providers/hcp"
      version = "0.0.1"
    }
  }
}

provider "hcp" {
  client_id     = "nop3"
  client_secret = ":D"
}

data "hcp_packer_image_iteration" "alpine" {
  bucket  = "ubuntu-xenial"
  channel = "production"
}

locals {
  images          = flatten(flatten(data.hcp_packer_image_iteration.alpine.builds[*].images[*]))
  image_us_east_1 = [for x in local.images: x if x.region == "us-east-1"][0]
  image_us_west_2 = [for x in local.images: x if x.region == "us-west-2"][0]
}

output "alpine-production-image_us_west_2" {
  value = local.image_us_west_2
}
```

I get the following output:

```shellsession
$ rm .terraform.lock.hcl; go build -o ./.terraform/providers/localhost/providers/hcp/0.0.1/darwin_amd64/ . && tf init && tf apply
[...]
Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

alpine-production-image_us_west_2 = {
  "created_at" = "2021-07-16T16:20:54.639Z"
  "id" = "01FAR1XV7FCFC04N725SA9ERE1"
  "image_id" = "ami-0ce9ac1ade7ce7fc9"
  "region" = "us-west-2"
}
```

This works ! But there's lots of HCL-fu and I kinda like the way I designed it in RFC 42 ! 🙂  Example from RFC 42:

```hcl
data "hcp_packer_image_iteration" "alpine" {
  image   = "hcp.dev/packer/images/${org}/${proj}/alpine"
  channel = "production"
 
  filter {
      nest_data = true # make of region a field
      name      = "region"
      values    = ["us_west_2", "us_east_1"]
  }
}

output "alpine-production-image_us_west_2" {
  value = data.hcp_packer_image_iteration.alpine.us_west_2
}

```

We have a few options: 
* leave it like this here and start developing something using the new Terraform plugin structure, this will allow me to return schemaless objects just like I want :D
* make this one here a bit better, do you have any suggestions ?
* a bit of both ?

Given time constrains, I would go for option 1.

File: docs/data-sources/packer_image_iteration.md
Comment (azr) [inline docs/data-sources/packer_image_iteration.md:13]: This is auto generated but, can I edit things here ?

File: internal/clients/hvn_route.go
Comment (azr) [reply internal/clients/hvn_route.go]: After updating the sdk, I could revert those changes.

**Comment by SwampDragons** [comment]
Yeah I think "leave this as is for the first round, and then implement something cleaner with the newer tools" feels like a very appropriate course of action

File: internal/provider/data_source_packer_image.go
Comment (azr) [reply internal/provider/data_source_packer_image.go]: So, after asking with the team: we kinda prefer to go with `bucket` as it is how it's being refered throughout in Packer hcp.

File: internal/provider/data_source_packer_image.go
Comment (azr) [reply internal/provider/data_source_packer_image.go]: It's done, tell me what you think; this one is very close to the schema of our API. 🙂

**Comment by azr** [comment]
Hey @bcmdarroch🙂, okay, I think I like the state of this code ! Tell me what you think !

Also, the acceptance tests work for me using my local dev or remote dev. But this feature is 'read-only' and the test require the data to be already stored on HCP. Is there a way for me to add some data to the HCP instance of the acc tests ? Or should the test create its own data too ?

File: docs/data-sources/packer_image_iteration.md
Comment (bcmdarroch) [reply docs/data-sources/packer_image_iteration.md:13]: To edit the generated doc, you'll want to add a `packer_image_iteration` template to https://github.com/hashicorp/terraform-provider-hcp/tree/main/templates/data-sources. You can follow the example of HVN route, and add whatever you need.

Follow-up question, is this feature generally available? If not, you'll want to add a banner in the template: `-> **Note:** This feature is currently in private/public beta.`

File: internal/provider/data_source_packer_image_iteration.go
Comment (bcmdarroch) [inline internal/provider/data_source_packer_image_iteration.go:52]: Is this ID guaranteed to be unique? I ask because Consul and Vault cluster IDs are unique only within a given project, so they use this [link helper](https://github.com/hashicorp/terraform-provider-hcp/blob/main/internal/provider/link.go#L71) to create a unique URL with org and project IDs.

Also wanted to note, so far we've set a pattern of reserving the ID field as an internal field for the provider, and using a prefix for the external ID (so something like `iteration_id`).

File: internal/provider/data_source_packer_image_iteration.go
Comment (bcmdarroch) [inline internal/provider/data_source_packer_image_iteration.go:135]:
```suggestion
							Description: "Status of this build. DONE means that all images tied to this build were successfully built.",
```

File: internal/provider/data_source_packer_image_iteration.go
Comment (bcmdarroch) [inline internal/provider/data_source_packer_image_iteration.go:189]: Ooo, could make this a shared helper 👀

**Review by bcmdarroch — COMMENTED** [review]
Looks good to me! This is definitely close. Just had an open question about the ID field. 

I also wasn't able to run the acceptance test (`[409] GetChannel default  &{Code:10 Details:[] Error:error querying bucket from db Message:error querying bucket from db}`), I'm guessing because it still needs a pre-config step that creates the image.

File: go.mod
Comment (azr) [reply go.mod:14]: Awesome, thanks !

File: internal/provider/data_source_packer_image_iteration.go
Comment (azr) [reply internal/provider/data_source_packer_image_iteration.go:52]: This iteration ID is the primary ID for iterations, yes, so it will be unique for HCP among all iterations ! 🙂. Not sure whether this is okay. I can rename it to `iteration_id`, to remain consistent, though !

**Comment by azr** [comment]
Thanks for reviewing, I still need to make the docs page look good and to fix the acc tests.

File: internal/provider/data_source_packer_image_iteration.go
Comment (azr) [reply internal/provider/data_source_packer_image_iteration.go:52]: I also used the `SetID` function with the same value, not sure if they are doing the same thing or not here ?

File: docs/data-sources/packer_image_iteration.md
Comment (azr) [reply docs/data-sources/packer_image_iteration.md:13]: It's private beta ! I added that, and tweaked things accordingly 👍🏼  thanks  !

Comment (azr) [reply docs/data-sources/packer_image_iteration.md:13]: is there a way to preview docs ?

Comment (azr) [reply docs/data-sources/packer_image_iteration.md]: Also, some nested attributes — like the `builds` one — are no documented, even though I did document the fields in the provider code; is there something I can do to fix that ?

File: internal/provider/data_source_packer_image_iteration.go
Comment (azr) [reply internal/provider/data_source_packer_image_iteration.go:52]: Actually, `SetID`, should be enough, removing this entry !

File: docs/data-sources/packer_image_iteration.md
Comment (bcmdarroch) [reply docs/data-sources/packer_image_iteration.md]: You can use this tool to preview docs, just have to copy-paste the markdown into the form: https://registry.terraform.io/tools/doc-preview

I'll check to see if the doc generator supports doubly nested objects. I'm thinking you'll need to add the `builds` documentation manually to the template in the meantime

File: go.mod
Comment (bcmdarroch) [inline go.mod:14]: Heads up, this just got bumped to v0.12.0 in https://github.com/hashicorp/terraform-provider-hcp/pull/178

Comment (azr) [reply go.mod:14]: Updating, thanks !

File: docs/data-sources/packer_image_iteration.md
Comment (azr) [reply docs/data-sources/packer_image_iteration.md]: > You can use this tool to preview docs, just have to copy-paste the markdown into the form: https://registry.terraform.io/tools/doc-preview

Dope, thanks ! Sorry I missed that info.

> The nested object fields get displayed below in a Nested Schema section:

Nice, also, I documented them in here:

https://github.com/hashicorp/terraform-provider-hcp/blob/7ae2cd98f817082540adc8234acef46f86e00c76/internal/provider/data_source_packer_image_iteration.go#L61-L76

But they are not documented here:

https://github.com/hashicorp/terraform-provider-hcp/blob/670702418fd8ac0837208c18d8a6f20a82b5cfc9/docs/data-sources/packer_image_iteration.md#L52-L59

**Comment by azr** [comment]
🥳  Super cool, acc tests pass ! Are we good with docs ? 🙂

File: internal/provider/data_source_packer_image_iteration.go
Comment (azr) [reply internal/provider/data_source_packer_image_iteration.go:184]: I'll maybe do that in another PR, if time !

**Comment by azr** [comment]
We're good with docs ! :shipit:

File: docs/data-sources/packer_image_iteration.md
Comment (azr) [reply docs/data-sources/packer_image_iteration.md]: So, I looked and long story short, because of the way TF schema works, this is not possible.

File: internal/provider/data_source_packer_image_iteration.go
Comment (bcmdarroch) [inline internal/provider/data_source_packer_image_iteration.go:26]: Are there any plans to make `bucket` and `channel` resources so this can all be managed by TF?

File: internal/provider/data_source_packer_image_iteration_test.go
Comment (bcmdarroch) [inline internal/provider/data_source_packer_image_iteration_test.go:176]: I've been getting an error when running this test:
```
{"error":"could not load the channel. something in the request is invalid. ","code":3,"message":"could not load the channel. something in the request is invalid. ","details":[]}
2021/08/05 12:05:49 [WARN] Got error running Terraform: exit status 1
```

I believe it's because the channel hasn't been upserted here.

File: internal/provider/data_source_packer_image_iteration_test.go
Comment (bcmdarroch) [inline internal/provider/data_source_packer_image_iteration_test.go:164]: Since this test upserts some test data that TF doesn't manage directly, we'll need an extra step to clean them up. You can add `CheckDestroy: testAccPackerCheckDestroy` here, then implement the CheckDestroy function to remove all the upserted data and verify the data source is deleted ([peering example](https://github.com/hashicorp/terraform-provider-hcp/blob/main/internal/provider/resource_aws_network_peering_test.go#L80))

**Review by bcmdarroch — COMMENTED** [review]
This is so close! Last step is to upsert one more dependency (`channel`) and then add a handler to clean up all the upserted dependencies.

**Review by bcmdarroch — APPROVED** [review]
Just tested with the fixes in https://github.com/hashicorp/terraform-provider-hcp/pull/180 and saw the acceptance test pass! 🎉 Let's get this in!

File: internal/provider/data_source_packer_image_iteration_test.go
Comment (nywilken) [reply internal/provider/data_source_packer_image_iteration_test.go:176]: @azr this issue has been address in #180 with the addition of the createChannel and updateChannel calls.

File: internal/provider/data_source_packer_image_iteration_test.go
Comment (nywilken) [reply internal/provider/data_source_packer_image_iteration_test.go:164]: @azr this has been implemented in #180 which @bcmdarroch has reviewed and approved. The change contain a few additional error checks, and a potential bug fix when requesting a channel that has not been assigned a complete iteration. 

I confirmed the panic using the provider on an exiting channel with no assigned iteration, and confirmed that the fix errors as expected. Merging the changes as they are for now so that we can get this in. But please follow up on the changes.

Comment (azr) [reply internal/provider/data_source_packer_image_iteration_test.go:320]: Thanks Wilken !

**Pull request merged**
